### PR TITLE
check for global uniqueness of case & form IDs

### PR DESCRIPTION
**Pull request by snopoke**
@dimagi/scale-team

### CODE REVIEW

File: corehq/form_processor/interfaces/processor.py [hunk 1]
```diff
@@ -191,24 +185,25 @@ def get_case_with_lock(self, case_id, lock=False, strip_history=False, wrap=Fals
...
             # case exists in other database
             raise IllegalCaseId("Bad case id")
 
         return case, lock
 
+    def other_db_processor(self):
+        """Get the processor for the database not used by this domain."""
+        from corehq.form_processor.backends.sql.processor import FormProcessorSQL
+        from corehq.form_processor.backends.couch.processor import FormProcessorCouch
+        (other_processor,) = {FormProcessorSQL, FormProcessorCouch} - {self.processor}
```
Comment (benrudolph) [inline corehq/form_processor/interfaces/processor.py]: nice was going to suggest this in the first commits

**Review by benrudolph — APPROVED** [review]
👍 this happens when we migrate a domain and then they submit a duplicate form right?

**Comment by benrudolph** [comment]
Or suppose it can happen with a new project space that is now in sql and uses the same case ids as another project

**Comment by snopoke** [comment]
The only time I've actually seen this is on projects that use the API and submit the same case / form ID's to multiple domains (one on couch and one on sql)

File: corehq/ex-submodules/casexml/apps/case/tests/test_domains.py [hunk 2]
```diff
@@ -2,13 +2,13 @@
...
 from corehq.apps.receiverwrapper.util import submit_form_locally
-from corehq.form_processor.tests.utils import run_with_all_backends
+from corehq.form_processor.tests.utils import run_with_all_backends, FormProcessorTestUtils
 
 ALICE_XML = """<?xml version='1.0' ?>
 <data xmlns:jrm="http://dev.commcarehq.org/jr/xforms" xmlns="http://openrosa.org/formdesigner/D95E58BD-A228-414F-83E6-EEE716F0B3AD">
     <name>Dinkle</name>
     <join_date>2010-08-28</join_date>
-    <n0:case case_id="ddb8e2b3-7ce0-43e4-ad45-d7a2eebe9169" user_id="user-xxx-alice" date_modified="2013-04-19T16:52:04.304-04" xmlns:n0="http://commcarehq.org/case/transaction/v2">
+    <n0:case case_id="ddb8e2b3-7ce0-43e4-ad45-domain-test" user_id="user-xxx-alice" date_modified="2013-04-19T16:52:04.304-04" xmlns:n0="http://commcarehq.org/case/transaction/v2">
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/case/tests/test_domains.py]: E501 line too long (180 > 115 characters)

File: corehq/ex-submodules/casexml/apps/case/tests/test_domains.py [hunk 3]
```diff
@@ -32,7 +32,7 @@
 EVE_XML = """<?xml version='1.0' ?>
 <data uiVersion="1" version="17" name="New Form" xmlns:jrm="http://dev.commcarehq.org/jr/xforms" xmlns="http://openrosa.org/formdesigner/1DFD8610-91E3-4409-BF8B-02D3B4FF3530">
     <plan_to_buy_gun>no</plan_to_buy_gun>
-    <n0:case case_id="ddb8e2b3-7ce0-43e4-ad45-d7a2eebe9169" user_id="user-xxx-eve" date_modified="2013-04-19T16:53:02.799-04" xmlns:n0="http://commcarehq.org/case/transaction/v2">
+    <n0:case case_id="ddb8e2b3-7ce0-43e4-ad45-domain-test" user_id="user-xxx-eve" date_modified="2013-04-19T16:53:02.799-04" xmlns:n0="http://commcarehq.org/case/transaction/v2">
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/case/tests/test_domains.py]: E501 line too long (178 > 115 characters)

File: corehq/ex-submodules/casexml/apps/case/tests/test_domains.py [hunk 4]
```diff
@@ -51,7 +51,7 @@
 ALICE_UPDATE_XML = """<?xml version='1.0' ?>
 <data uiVersion="1" version="17" name="New Form" xmlns:jrm="http://dev.commcarehq.org/jr/xforms" xmlns="http://openrosa.org/formdesigner/1DFD8610-91E3-4409-BF8B-02D3B4FF3530">
     <plan_to_buy_gun>no</plan_to_buy_gun>
-    <n0:case case_id="ddb8e2b3-7ce0-43e4-ad45-d7a2eebe9169" user_id="user-xxx-alice" date_modified="2013-04-19T16:53:02.799-04" xmlns:n0="http://commcarehq.org/case/transaction/v2">
+    <n0:case case_id="ddb8e2b3-7ce0-43e4-ad45-domain-test" user_id="user-xxx-alice" date_modified="2013-04-19T16:53:02.799-04" xmlns:n0="http://commcarehq.org/case/transaction/v2">
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/case/tests/test_domains.py]: E501 line too long (180 > 115 characters)

**Review by emord — APPROVED** [review]

File: corehq/ex-submodules/casexml/apps/case/tests/test_domains.py [hunk 5]
```diff
@@ -75,8 +75,13 @@
 @override_settings(CASEXML_FORCE_DOMAIN_CHECK=True)
 class DomainTest(TestCase):
 
+    def tearDown(self):
+        FormProcessorTestUtils.delete_all_cases_forms_ledgers()
+        super(DomainTest, self).tearDown()
+
+    @run_with_all_backends
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/case/tests/test_domains.py]: F821 undefined name 'run_with_all_backends'

**Pull request merged**